aclk/agent_cloud_link.c

```diff
@@ -1443,75 +1481,174 @@ int aclk_update_alarm(RRDHOST *host, ALARM_ENTRY *ae)
+        error("The \"version\" message came too late ignoring.");
+        goto err_cleanup;
+    }
+    if (unlikely(aclk_shared_state.version_neg)) {
+        errno = 0;
+        error("Version has already been set to %d", aclk_shared_state.version_neg);
+        goto err_cleanup;
     }
+    aclk_shared_state.version_neg = version;
     ACLK_SHARED_STATE_UNLOCK;
 
+    info("Choosing version %d of ACLK", version);
+
+    return 0;
+
+err_cleanup:
+    ACLK_SHARED_STATE_UNLOCK;
+    return 1;
+}
+
+struct {
+    char *name;
+    int(*fnc)(struct aclk_request *cloud_to_agent);
+} aclk_incoming_msg_types[] = {
+    { .name = "http",    .fnc = aclk_handle_cloud_request    },
+    { .name = "version", .fnc = aclk_handle_version_response },
+    { .name = NULL,      .fnc = NULL                         }
+};
+
+int aclk_handle_cloud_message(char *payload)
+{
+    struct aclk_request cloud_to_agent;
+    memset(&cloud_to_agent, 0, sizeof(struct aclk_request));
+
+    if (aclk_stats_enabled) {
+        ACLK_STATS_LOCK;
+        aclk_metrics_per_sample.cloud_req_recvd++;
+        ACLK_STATS_UNLOCK;
+    }
+
     if (unlikely(!payload)) {
         debug(D_ACLK, "ACLK incoming message is empty");
-        return 0;
+        goto err_cleanup;
     }
 
     debug(D_ACLK, "ACLK incoming message (%s)", payload);
 
     int rc = json_parse(payload, &cloud_to_agent, cloud_to_agent_parse);
 
-    if (unlikely(
-            JSON_OK != rc || !cloud_to_agent.payload || !cloud_to_agent.callback_topic || !cloud_to_agent.msg_id ||
-            !cloud_to_agent.type_id || cloud_to_agent.version > ACLK_VERSION ||
-            strcmp(cloud_to_agent.type_id, "http"))) {
-        if (JSON_OK != rc)
-            error("Malformed json request (%s)", payload);
-
-        if (cloud_to_agent.version > ACLK_VERSION)
-            error("Unsupported version in JSON request %d", cloud_to_agent.version);
-
-        if (cloud_to_agent.payload)
-            freez(cloud_to_agent.payload);
-
-        if (cloud_to_agent.type_id)
-            freez(cloud_to_agent.type_id);
-
-        if (cloud_to_agent.msg_id)
-            freez(cloud_to_agent.msg_id);
+    if (unlikely(rc != JSON_OK)) {
+        errno = 0;
+        error("Malformed json request (%s)", payload);
+        goto err_cleanup;
+    }
 
-        if (cloud_to_agent.callback_topic)
-            freez(cloud_to_agent.callback_topic);
+    if (!cloud_to_agent.type_id) {
+        errno = 0;
+        error("Cloud message is missing compulsory key \"type\"");
+        goto err_cleanup;
+    }
 
-        if (aclk_stats_enabled) {
-            ACLK_STATS_LOCK;
-            aclk_metrics_per_sample.cloud_req_err++;
-            ACLK_STATS_UNLOCK;
+    for (int i = 0; aclk_incoming_msg_types[i].name; i++) {
+        if (strcmp(cloud_to_agent.type_id, aclk_incoming_msg_types[i].name) == 0) {
+            if (likely(!aclk_incoming_msg_types[i].fnc(&cloud_to_agent))) {
+                freez(cloud_to_agent.type_id);
+                return 0;
```

Comment:
I think this will leak memory (not going through err_cleanup)

Reply:
the handlers are supposed to clean after themselves. in case of cloud request, the `aclk_queue_query` will take ownership of pointers so we cannot free it! this was how it was before this PR

Reply:
This is done to avoid need to copy the strings when we add them to queue!

Reply:
What I can do though is to explicitly document this. I think that is good idea